### PR TITLE
Workflow_monitoring

### PR DESCRIPTION
**Pull request by michaelmattig**
- [ ] I added an entry to [`CHANGELOG.md`](CHANGELOG.md) if knowledge of this change could be valuable to users.

---

Here is a brief summary of what I did:

<TEXT>

### CODE REVIEW

File: operators/src/processing/reprojection.rs [hunk 1]
```diff
@@ -275,13 +276,22 @@ impl RasterOperator for Reprojection {
...
+        Ok(context.initialize_operator(op.boxed(), Box::new(ReprojectionSpan {})))
+    }
+}
+
+#[derive(Clone)]
+struct ReprojectionSpan {}
+
+impl CreateSpan for ReprojectionSpan {
+    fn create_span(&self) -> Span {
+        span!(Level::TRACE, "Reprojection")
```
Comment (ChristianBeilschmidt) [inline operators/src/processing/reprojection.rs:294]: Maybe we should create a macro for that.

File: operators/src/util/raster_stream_to_png.rs [hunk 2]
```diff
@@ -23,6 +24,10 @@ pub async fn raster_stream_to_png_bytes<T, C: QueryContext>(
 where
     T: Pixel,
 {
+    // TODO: rather use `span.in_scope` because it is safer inside async functions (cf. https://docs.rs/tracing/latest/tracing/span/struct.Span.html#method.enter)
+    let span = span!(Level::TRACE, "raster_stream_to_png_bytes");
+    let _enter = span.enter();
```
Comment (ChristianBeilschmidt) [inline operators/src/util/raster_stream_to_png.rs:29]: can we use in_scope here?

File: services/src/bin/main.rs [hunk 3]
```diff
@@ -58,7 +96,7 @@ fn initialize_logging() -> Option<FileLogWriterHandle> {
         // Thus, we use UTC time instead.
         flexi_logger::DeferredNow::force_utc();
 
-        let (file_writer, fw_handle) = FileLogWriter::builder(file_spec)
+        let (file_writer, _fw_handle) = FileLogWriter::builder(file_spec)
```
Comment (ChristianBeilschmidt) [inline services/src/bin/main.rs:99]: what changed around here?

File: services/src/error.rs [hunk 4]
```diff
@@ -56,6 +56,9 @@ pub enum Error {
     Proj {
         source: proj::ProjError,
     },
+    Trace {
+        source: opentelemetry::trace::TraceError,
+    },
```
Comment (ChristianBeilschmidt) [inline services/src/error.rs:61]: Can't we wrap that into an Opentelemetry error?

File: services/src/error.rs [hunk 5]
```diff
@@ -521,3 +524,9 @@ impl From<tokio::task::JoinError> for Error {
         Error::TokioJoin { source }
     }
 }
+
+impl From<opentelemetry::trace::TraceError> for Error {
+    fn from(source: opentelemetry::trace::TraceError) -> Self {
+        Error::Trace { source }
+    }
+}
```
Comment (ChristianBeilschmidt) [inline services/src/error.rs:532]: Use `#[snafu(context(false))]` instead

File: services/src/pro/contexts/in_memory.rs [hunk 6]
```diff
@@ -22,6 +22,8 @@ use snafu::ResultExt;
 use std::path::PathBuf;
 use std::sync::Arc;
```
Comment (ChristianBeilschmidt) [inline services/src/pro/contexts/in_memory.rs:24]:
```suggestion
```

File: services/src/pro/contexts/mod.rs [hunk 7]
```diff
@@ -24,3 +37,202 @@ pub trait ProContext: Context<Session = UserSession> {
...
+    }
+
+    fn initialize_operator(
+        &self,
+        op: Box<dyn geoengine_operators::engine::InitializedRasterOperator>,
+        span: Box<dyn geoengine_operators::engine::CreateSpan>,
+    ) -> Box<dyn geoengine_operators::engine::InitializedRasterOperator> {
+        Box::new(InitializedProcessorStatistics::statistics_with_id(
+            op,
+            "gdal_source".to_string(),
```
Comment (ChristianBeilschmidt) [inline services/src/pro/contexts/mod.rs:105]: why is everything a gdal source?

File: services/src/pro/contexts/mod.rs [hunk 8]
```diff
@@ -24,3 +37,202 @@ pub trait ProContext: Context<Session = UserSession> {
...
+    D: DatasetDb<S>,
+    L: LayerProviderDb,
+    S: Session,
+{
+    dataset_db: Arc<D>,
+    layer_provider_db: Arc<L>,
+    thread_pool: Arc<ThreadPool>,
+    session: S,
+    tiling_specification: TilingSpecification,
+}
```
Comment (ChristianBeilschmidt) [inline services/src/pro/contexts/mod.rs:52]: Why did you need to create that?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: Settings-default.toml [hunk 9]
```diff
@@ -110,3 +110,7 @@ issuer = ""
 client_id = ""
 redirect_uri = ""
 scopes = [ ]
+
+[open_telemetry]
+enabled = false
+endpoint = "127.0.0.1:6831"
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply Settings-default.toml:116]: this is the address of the JAEGER instance where Geo Engine sends the traces to. It might be even on a different machine so we need the configuration

File: operators/src/engine/execution_context.rs [hunk 10]
```diff
@@ -28,6 +28,12 @@ pub trait ExecutionContext: Send
 {
     fn thread_pool(&self) -> &Arc<ThreadPool>;
     fn tiling_specification(&self) -> TilingSpecification;
+
+    fn initialize_operator(
+        &self,
+        op: Box<dyn InitializedRasterOperator>,
+        span: Box<dyn CreateSpan>,
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/engine/execution_context.rs:35]: I can't make the function generic over the `CreateSpan` because then it can no longer be made into a trait object.

File: operators/src/pro/meta/statistics.rs [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    pub fn statistics_with_id(source: S, id: String, span: Box<dyn CreateSpan>) -> Self {
+        Self { source, id, span }
+    }
+}
+
+impl InitializedRasterOperator
+    for InitializedProcessorStatistics<Box<dyn InitializedRasterOperator>>
+{
+    fn result_descriptor(&self) -> &RasterResultDescriptor {
+        debug!("[{}] | raster result descriptor", self.id);
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/pro/meta/statistics.rs:30]: I guess we have to decide in general, what is a log message and what is a tracing event. It's possible to convert one thing into the other but they are not the same. E.g. the swagger UI URL log message should most likely not be a tracing event.

File: services/src/pro/contexts/mod.rs (hunk 8 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs:52]: the execution context is now responsible for wrapping the operators with a statistics operator. As it should only do so in the pro version, we need a separate ExecutionContextImpl for Pro. Previously we used the same in pro and free.

File: services/src/error.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/error.rs:532]: why? maybe we want to see why the tracing failed.

File: operators/src/processing/reprojection.rs (hunk 1 above)
Comment (michaelmattig) [reply operators/src/processing/reprojection.rs:294]: if we decide that this is the way to go for span creation we should.

File: operators/src/pro/adapters/stream_statistics_adapter.rs [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+            "[{}] | poll_next_count: {}",
+            *this.id,
+            *this.poll_next_count
+        );
+
+        let v = ready!(this.stream.as_mut().poll_next(cx));
+        match v {
+            Some(_) => {
+                *this.element_count += 1;
+                debug!(
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/pro/adapters/stream_statistics_adapter.rs:70]: it's a judgement call. I used Johannes classification of the log events here. You can suggest a different one if you don't like it.

File: operators/src/pro/adapters/stream_statistics_adapter.rs [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        let mut this = self.project();
+        *this.poll_next_count += 1;
+
+        let _enter = this.span.enter();
+
+        trace!(
+            "[{}] | poll_next_count: {}",
+            *this.id,
+            *this.poll_next_count
+        );
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/pro/adapters/stream_statistics_adapter.rs]: 60deff1f01a061f9903fafef13cdf967b4f0cc08

File: operators/src/pro/adapters/stream_statistics_adapter.rs [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+                debug!(
+                    "[{}] | poll_next_count: {} | element_count: {} | next",
+                    *this.id, *this.poll_next_count, *this.element_count
+                );
+            }
+            None => {
+                debug!(
+                    "[{}] | poll_next_count: {} | element_count: {} | empty",
+                    *this.id, *this.poll_next_count, *this.element_count
+                );
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/pro/adapters/stream_statistics_adapter.rs]: 60deff1f01a061f9903fafef13cdf967b4f0cc08

File: operators/src/pro/meta/statistics.rs [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+        match processor_result {
+            Ok(p) => {
+                let res_processor = match p {
+                    TypedRasterQueryProcessor::U8(p) => TypedRasterQueryProcessor::U8(Box::new(
+                        ProcessorStatisticsProcessor::statistics_with_id(
+                            p,
+                            self.id.clone(),
+                            self.span.clone(),
+                        ),
+                    )),
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/pro/meta/statistics.rs:45]: I don't think so. Neither call_generic_raster_processor nor call_on_generic_raster_processor helps me here. Or am I missing something? Please make a concrete suggestion.

File: operators/src/util/raster_stream_to_png.rs (hunk 2 above)
Comment (michaelmattig) [reply operators/src/util/raster_stream_to_png.rs:29]: yes but this would either mean wrapping the whole function body inside a closure or introduce another method, that's why I didn't do it yet.

File: services/src/bin/main.rs [hunk 16]
```diff
@@ -1,55 +1,93 @@
...
+            tracing_opentelemetry::layer().with_tracer(tracer);
+        Some(opentelemetry)
+    } else {
+        None
+    };
+
+    let registry = registry.with(open_telemetry_layer);
+    let registry = registry.with(console_layer(&logging_config));
+    let registry = registry.with(file_layer(&logging_config));
+    registry.init();
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply services/src/bin/main.rs:57]: 1. what do you mean? this is already a function with the whole purpose of setting up the logging and then calling the actual start method. If I extract it into another method it would be exactly the same.
2. this is not a service but a global logging/tracing subscriber. How does this to relate to actix services and routing?

File: services/src/bin/main.rs (hunk 3 above)
Comment (michaelmattig) [reply services/src/bin/main.rs:99]: nothing. I just ignore the handle at a different location in the code. We didn't use it before too

File: services/src/error.rs (hunk 4 above)
Comment (michaelmattig) [reply services/src/error.rs:61]: This IS the opentelemetry error.

File: services/src/pro/contexts/in_memory.rs (hunk 6 above)
Comment (michaelmattig) [reply services/src/pro/contexts/in_memory.rs:24]: ?

File: services/src/pro/contexts/mod.rs (hunk 7 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs:105]: yeah this is wrong. I could add a `get_id` method to `CreateSpan` to set it here, or just get rid of the "id" here. It's somewhat redundant because it would always be the same for every span.

File: operators/src/pro/meta/statistics.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply operators/src/pro/meta/statistics.rs:45]: `call_generic_raster_processor` should be it… why doesn't it work here?

File: services/src/error.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply services/src/error.rs:61]: It is a trace error, right?
Aren't there other errors with respect to our spectacular remote logging?

File: services/src/error.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply services/src/error.rs:532]: But you are implementing exactly what context(false) auto-implements

File: operators/src/engine/execution_context.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/engine/execution_context.rs]: I changed some stuff and it is no longer boxed 68617472345bdacfa1a856184f118722e110b5b3

File: operators/src/pro/meta/statistics.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/pro/meta/statistics.rs]: because we `match` on a `TypedRasterQueryProcessor` and the macro matches a `RasterDataType`

File: services/src/error.rs (hunk 4 above)
Comment (michaelmattig) [reply services/src/error.rs:61]: I don't know if it is `spectacular` but remote logging is called "tracing" and the `trace` module of opentelemetry is the "API for tracing applications and libraries." and thus the ` TraceError` should be exactly what we need.

File: services/src/error.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/error.rs]: ed702fe821950f56ee1a6ab0d56aacc25adbf96c

File: services/src/pro/contexts/mod.rs (hunk 7 above)
Comment (michaelmattig) [reply services/src/pro/contexts/mod.rs]: I removed the id for now because it was exactly the same as the span, thus carried no information. 68617472345bdacfa1a856184f118722e110b5b3

File: operators/src/pro/meta/statistics.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply operators/src/pro/meta/statistics.rs]: We have no macro from TypedRasterQueryProcessor to TypedRasterQueryProcessor?
Maybe you can create one?

File: operators/src/mock/mock_dataset_data_source.rs [hunk 17]
```diff
@@ -130,6 +131,10 @@ impl VectorOperator for MockDatasetDataSource {
         }
         .boxed())
     }
+
+    fn span(&self) -> CreateSpan {
+        || span!(Level::TRACE, MockDatasetDataSource::TYPE_NAME)
+    }
```
Comment (ChristianBeilschmidt) [inline operators/src/mock/mock_dataset_data_source.rs:137]: Is there any way we can use the `OperatorName` trait for that?

File: operators/src/plot/histogram.rs [hunk 18]
```diff
@@ -141,6 +146,10 @@ impl PlotOperator for Histogram {
             }
         })
     }
+
+    fn span(&self) -> CreateSpan {
+        || span!(Level::TRACE, Histogram::TYPE_NAME)
+    }
```
Comment (ChristianBeilschmidt) [inline operators/src/plot/histogram.rs:152]: Maybe we can at least use a macro for this `fn`. This way, we can easily change the behavior for all operators.

File: operators/src/mock/mock_dataset_data_source.rs (hunk 17 above)
Comment (michaelmattig) [reply operators/src/mock/mock_dataset_data_source.rs:137]: I do use the `OperatorName` trait, as the  `TYPE_NAME` field is part of it. However, like I tried to explain in the Trello card, we can't use the trait as a generic. We can only input the `TYPE_NAME` of a concrete type. Otherwise Rust complains on not being able to caputre the outer generic of a function. This has to do with the way the `span!` macro works, as far as I undrstand.

File: operators/src/mock/mock_feature_collection_source.rs [hunk 19]
```diff
@@ -83,6 +85,10 @@ where
 
 pub type MockFeatureCollectionSource<G> = SourceOperator<MockFeatureCollectionSourceParams<G>>;
 
+impl<G: Geometry + ArrowTyped> OperatorName for MockFeatureCollectionSource<G> {
+    const TYPE_NAME: &'static str = "MockFeatureCollectionSource";
+}
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/mock/mock_feature_collection_source.rs:90]: yes, this is an *impl*ementation of that trait, see line 88

File: operators/src/engine/operator.rs [hunk 20]
```diff
@@ -28,18 +29,30 @@ pub trait OperatorData {
 pub trait RasterOperator:
     CloneableRasterOperator + OperatorData + Send + Sync + std::fmt::Debug
 {
-    async fn initialize(
+    // internal logic of the operator
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/engine/operator.rs]: 11c4d2df4f4ea9e34f970b7bebfb1f7dfd2b680a

File: operators/src/engine/mod.rs [hunk 21]
```diff
@@ -168,25 +168,4 @@ macro_rules! call_on_bi_generic_raster_processor {
...
-        Box::new(self.clone())
-    }
-}
-
-impl Clone for Box<dyn CreateSpan> {
-    fn clone(&self) -> Box<dyn CreateSpan> {
-        self.clone_create_span()
-    }
-}
+pub type CreateSpan = fn() -> Span;
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply operators/src/engine/mod.rs:172]: 11c4d2df4f4ea9e34f970b7bebfb1f7dfd2b680a

**Comment by michaelmattig** [comment]
> Please also write a changelog.

&

> I added an entry to [CHANGELOG.md](https://github.com/geo-engine/geoengine/pull/CHANGELOG.md) if knowledge of this change could be valuable to users

knowledge of monitoring is not valuable to users so I didn't add it as to not bloat the changelog

File: operators/src/mock/mock_dataset_data_source.rs (hunk 17 above)
Comment (ChristianBeilschmidt) [reply operators/src/mock/mock_dataset_data_source.rs:137]: Is this the issue?
https://github.com/tokio-rs/tracing/issues/1047

File: operators/src/plot/histogram.rs (hunk 18 above)
Comment (ChristianBeilschmidt) [reply operators/src/plot/histogram.rs:152]: Would be nice, but we could have a normal macro instead here.

File: operators/src/mock/mock_dataset_data_source.rs (hunk 17 above)
Comment (michaelmattig) [reply operators/src/mock/mock_dataset_data_source.rs:137]: no, the `'static` is not the problem, but the string having to be `const` in order for the macro to work

File: operators/src/plot/histogram.rs (hunk 18 above)
Comment (michaelmattig) [reply operators/src/plot/histogram.rs]: de3d837b15aa93ceeaa690437b8747d2d5457067

**Review by ChristianBeilschmidt — APPROVED** [review]
